### PR TITLE
Add support for the SWP instruction and the AtomicReadAndWriteOp to AArch64

### PR DESCRIPTION
**Pull request by daniel-stewart-qdt**
These patches add support for the SWP instruction and also implement the AtomicReadAndWritOp LIR instruction for AArch64. This is in preparation for the implementation of getAndSetInt() and getAndSetLong() intrinsics.

### CODE REVIEW

**Comment by axel22** [comment]
@daniel-stewart-qdt FYI, the JDK8-style `compareAndSwap` and `compareAndExchange` intrinsics for sub-32-bits data types were recently implemented for Amd64, but the ones for Aarch64 are currently missing. The hooks for adding these instructions to the Aarch64 backed are here:

https://github.com/oracle/graal/commit/55045bff8eacaa4bd003c90e8b040dd908d10afa#diff-6fff1920efa19ddf2e0bc614a05c4351R60

https://github.com/oracle/graal/commit/55045bff8eacaa4bd003c90e8b040dd908d10afa#diff-366e04a3d2a82a9c990b3ad549f90a32R130

I planned to do this at some point in the future, but if you want to take a stab at this, you're more than welcome.

And thanks for contributing!

P.S. I am not an expert on Aarch64, but I thought that the `SWP` instruction was deprecated in favour of `LDREX` and `STREX`?

**Comment by daniel-stewart-qdt** [comment]
@axel22 Thanks for the heads up about the sub-32 bit implementations. I'll try and go ahead and implement those for AArch64. 

You are correct that the SWP instruction was deprecated from ARMv6 and on. But from [here](https://developer.arm.com/products/architecture/a-profile/docs/dht0008/latest/a-swp-and-swpb/a1-legacy-synchronization-instructions), that appears to be due to "The SWP and SWPB instructions were used for synchronization before the introduction of exclusive accesses into the ARM architecture."

The SWP instruction for AArch64 was added as an 8.1 extension, at the same time as the CAS and LDADD instructions, and after the LDREX/STREX instructions had already been defined. I believe this is now the preferred way to do atomic operations, if your architecture implements 8.1 and above.

File: compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java [hunk 1]
```diff
@@ -154,4 +154,55 @@ public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
...
+        @Override
+        public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
+            assert accessKind.isInteger();
+            final int size = accessKind.getSizeInBytes() * Byte.SIZE;
+
+            Register address = asRegister(addressValue);
+            Register value = asRegister(newValue);
+            Register result = asRegister(resultValue);
+
+            if (AArch64LIRFlagsVersioned.useLSE(masm)) {
```
Comment (sanzinger) [inline compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java]: FYI, I have consolidated the different implementations.

**Comment by axel22** [comment]
@daniel-stewart-qdt I might be wrong, but I think you need to enable the `getAndSet` intrinsification for AArch64 in the `AArch64GraphBuilderPlugins` class (see `AMD64GraphBuilderPlugins`).

**Comment by axel22** [comment]
Also, could you please enable the `testGetAndSet` method `UnsafeReplacementsTest`.

**Comment by sanzinger** [comment]
Thx for the comment. I'll add it.

**Comment by daniel-stewart-qdt** [comment]
@axel22 Yes you are correct that the getAndSet portion was not implemented. I tried to indicate that in the original comments, but perhaps it wasn't clear. I was waiting for the previous getAndAdd code to go in so that I could leverage some of that code, since there is common code there. In the meantime, I see that @sanzinger has done some nice refactoring that definitely makes the code better.

**Comment by sanzinger** [comment]
Heads up! I updated your branch seconds ago.

**Comment by daniel-stewart-qdt** [comment]
Thanks! I still think newValue in AtomicReadAndWriteOp needs to be an AllocatableValue instead of just a Value. Given what you are doing, are you planning to update that or should I push a new patch which incorporates what you've done?

**Comment by sanzinger** [comment]
Yes, the newValue must be an allocatable. Please update accordingly. The current change also fails in the assembler, since it also tries to deal with sub-int types:
```
$ mx unittest UnsafeReplacementsTest
testGetAndAdd(org.graalvm.compiler.replacements.jdk9.UnsafeReplacementsTest)
org.graalvm.compiler.debug.GraalError: java.lang.AssertionError
        at lir instruction: B3@94 org.graalvm.compiler.lir.aarch64.AArch64AtomicMove$AtomicReadAndAddOp r2|BYTE = ATOMIC_READ_AND_ADD (~addressValue: r0|QWORD
[.], ~deltaValue: r1|DWORD) {scratchValue1: r3|BYTE, scratchValue2: r4|BYTE} accessKind: BYTE
[B0, B2, B3, B1]
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.asm.CompilationResultBuilder.emitOp(CompilationResultBuilder.java:540)
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.asm.CompilationResultBuilder.emitBlock(CompilationResultBuilder.java:522)
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.asm.CompilationResultBuilder.emit(CompilationResultBuilder.java:497)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.aarch64.AArch64HotSpotBackend.emitCodeBody(AArch64HotSpotBackend.java:317)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.aarch64.AArch64HotSpotBackend.emitCode(AArch64HotSpotBackend.java:260)
        at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.emitCode(GraalCompiler.java:394)
        at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.emitBackEnd(GraalCompiler.java:288)
        at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:181)
        at org.graalvm.compiler.core.test.GraalCompilerTest.compile(GraalCompilerTest.java:1060)
        at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:957)
        at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:898)
        at org.graalvm.compiler.core.test.GraalCompilerTest.executeActual(GraalCompilerTest.java:734)
        at org.graalvm.compiler.core.test.GraalCompilerTest.executeActualCheckDeopt(GraalCompilerTest.java:866)
        at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:856)
        at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:852)
        at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:821)
        at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:790)
        at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:783)
        at org.graalvm.compiler.replacements.jdk9.UnsafeReplacementsTest.testGetAndAdd(UnsafeReplacementsTest.java:256)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:280)
        at com.oracle.mxtool.junit.MxJUnitWrapper.main(MxJUnitWrapper.java:196)
Caused by: java.lang.AssertionError
        at jdk.internal.vm.compiler/org.graalvm.compiler.asm.aarch64.AArch64Assembler$InstructionType.generalFromSize(AArch64Assembler.java:393)
        at jdk.internal.vm.compiler/org.graalvm.compiler.asm.aarch64.AArch64Assembler.add(AArch64Assembler.java:1764)
        at jdk.internal.vm.compiler/org.graalvm.compiler.asm.aarch64.AArch64MacroAssembler.add(AArch64MacroAssembler.java:523)
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.aarch64.AArch64AtomicMove$AtomicReadAndAddOp.emitCode(AArch64AtomicMove.java:150)
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.aarch64.AArch64LIRInstruction.emitCode(AArch64LIRInstruction.java:39)
        at jdk.internal.vm.compiler/org.graalvm.compiler.lir.asm.CompilationResultBuilder.emitOp(CompilationResultBuilder.java:535)
        ... 48 more
```

**Comment by sanzinger** [comment]
Thank you very much for the fix!

**Pull request merged**